cdap-hbase-compat-0.98/src/main/java/co/cask/cdap/data2/transaction/messaging/coprocessor/hbase98/MessageTableRegionObserver.java

```diff
@@ -0,0 +1,162 @@
+      metadataTable = env.getTable(nameConverter.toTableName(hbaseNamespacePrefix,
+                                                             TableId.from(metadataTableNamespace, metadataTableName)));
+      topicCache = CacheBuilder.newBuilder()
+        .expireAfterWrite(2, TimeUnit.MINUTES)
+        .maximumSize(1000)
+        .build(new CacheLoader<TopicId, TopicMetadata>() {
+
+          @Override
+          public TopicMetadata load(TopicId topicId) throws Exception {
+            Get get = new Get(MessagingUtils.toRowKeyPrefix(topicId));
+            Result result = metadataTable.get(get);
+            byte[] properties = result.getValue(COL_FAMILY, COL);
+            Map<String, String> propertyMap = GSON.fromJson(Bytes.toString(properties), MAP_TYPE);
+            return new TopicMetadata(topicId, propertyMap);
+          }
+        });
+    }
+  }
+
+  @Override
+  public InternalScanner preFlushScannerOpen(ObserverContext<RegionCoprocessorEnvironment> c, Store store,
+                                             KeyValueScanner memstoreScanner, InternalScanner s) throws IOException {
+    LOG.info("preFlush, filter using TTLFilter");
+    Scan scan = new Scan();
+    scan.setFilter(new TTLFilter(c.getEnvironment(), System.currentTimeMillis()));
+    return new StoreScanner(store, store.getScanInfo(), scan, Collections.singletonList(memstoreScanner),
+                            ScanType.COMPACT_DROP_DELETES, store.getSmallestReadPoint(), HConstants.OLDEST_TIMESTAMP);
+  }
+
+  @Override
+  public InternalScanner preCompactScannerOpen(ObserverContext<RegionCoprocessorEnvironment> c, Store store,
+                                               List<? extends KeyValueScanner> scanners, ScanType scanType,
+                                               long earliestPutTs, InternalScanner s,
+                                               CompactionRequest request) throws IOException {
+    LOG.info("preCompact, filter using TTLFilter");
+    Scan scan = new Scan();
+    scan.setFilter(new TTLFilter(c.getEnvironment(), System.currentTimeMillis()));
+    return new StoreScanner(store, store.getScanInfo(), scan, scanners, scanType, store.getSmallestReadPoint(),
+                            earliestPutTs);
+  }
+
+  private final class TTLFilter extends FilterBase {
+    private final RegionCoprocessorEnvironment env;
+    private final long timestamp;
+
+    TTLFilter(RegionCoprocessorEnvironment env, long timestamp) {
+      this.env = env;
+      this.timestamp = timestamp;
+    }
+
+    @Override
+    public ReturnCode filterKeyValue(Cell cell) throws IOException {
+      byte[] rowKey = Arrays.copyOfRange(cell.getRow(), prefixLength, cell.getRowLength());
+      ImmutableMessageTableEntry entry = new ImmutableMessageTableEntry(rowKey, null, null);
+      try {
+        TopicMetadata metadata = topicCache.get(entry.getTopicId());
+        if ((timestamp - entry.getPublishTimestamp()) > TimeUnit.SECONDS.toMillis(metadata.getTTL())) {
```

Comment:
The TTL conversion doesn't need to happen in here. You can have the cache returns the TTL in ms so that in here you just lookup. Also, because this filter is used in scan, meaning you don't have to lookup the ttl from topic for each cell. You only need to lookup when the topic changed.